libs/models/datapoints.py

```diff
@@ -0,0 +1,400 @@
+        forecasted: bool = False,
+    ) -> Optional["Consumption"]:
+        try:
+            return Consumption(
+                zoneKey=zone_key,
+                datetime=datetime,
+                source=source,
+                consumption=consumption,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(
+                f"Error creating consumption datapoint {datetime}: {e}",
+                extra={
+                    "zone_key": zone_key,
+                    "datetime": datetime,
+                    "kind": "consumption",
+                },
+            )
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "consumption": self.consumption,
+            "source": self.source,
+        }
+
+
+class Price(Datapoint):
+    price: float
+    currency: str
+
+    @validator("currency")
+    def _validate_currency(cls, v: str):
+        if v not in VALID_CURRENCIES:
+            raise ValueError(f"Unknown currency: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        price: float,
+        currency: str,
+        forecasted: bool = False,
+    ) -> Optional["Price"]:
+        try:
+            return Price(
+                zoneKey=zone_key,
+                datetime=datetime,
+                source=source,
+                price=price,
+                currency=currency,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating price datapoint {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "currency": self.currency,
+            "price": self.price,
+            "source": self.source,
+        }
+
+
+class DatapointBatch(ABC):
+    """A wrapper around datapoints lists."""
+
+    logger: Logger
+    datapoints: List[Datapoint]
+
+    def __init__(self, logger: Logger):
+        self.datapoints = list()
+        self.logger = logger
+
+    @abstractmethod
+    def append(self, **kwargs):
+        """Handles creation of datapoints and adding it to the batch."""
+        # TODO Handle one day the creation of mixed batches.
+        pass
+
+    def to_dict(self) -> List[Dict[str, Any]]:
+        return [datapoint.to_dict() for datapoint in self.datapoints]
+
+
+class ExchangeBatch(DatapointBatch):
+    def append(
+        self,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        value: float,
+        forecasted: bool = False,
+    ):
+        datapoint = Exchange.create(
+            self.logger, zone_key, datetime, source, value, forecasted
+        )
+        if datapoint:
+            self.datapoints.append(datapoint)
```

Comment:
should we log which data points were excluded or will this overload our logs?

Reply:
We are doing it, inside the create method of each element errors are logged. I think it is important to do and would not overload us once we add the improved logging that I started.